### ISSUE TITLE
Timed out fetching a new connection from the pool

### ISSUE THREAD

**Issue by P4sca1** [prisma/prisma#7249]
### Bug description

After running my application for around 1-2 days, I start getting the following error:
```
prisma:client Error: Timed out fetching a new connection from the pool. Please consider reducing the number of requests or increasing the `connection_limit` parameter (https://www.prisma.io/docs/concepts/components/prisma-client/connection-management#connection-pool). Current limit: 10.
```
The same happened before with a connection limit of 3. Raising it to 10 seemed to have no effect.

After the error occurs once, no query will succeed anymore, the error will be thrown on any query and the application requires a restart (prisma will never recover from the error).

I checked whether the database has an issue, but it is working fine.
`SHOW PROCESSLIST` shows that **no** connection is actually established.

### How to reproduce

I have not found a way to reliably reproduce this yet. I always need to wait 1-2 days until the error randomly appears.
I am running my application in debug mode with `DEBUG=*` for the last days, so if you need any additional information, I can look it up in the logs. 

### Prisma information

```
generator client {
    provider        = "prisma-client-js"
    binaryTargets   = ["native"]
}

datasource db {
    provider = "mysql"
    url      = env("DATABASE_URL")
}
```

I can't post my complete prisma schema here.
It doesn't matter which queries are run, all of them fail with the same error.

### Environment & setup

- OS: Linux (official node:14 docker image)
- Database: MySQL 8.0
- Node.js version: 14

### Prisma Version

```
prisma               : 2.21.2
@prisma/client       : 2.21.2
Current platform     : darwin
Query Engine         : query-engine e421996c87d5f3c8f7eeadd502d4ad402c89464d (at node_modules/prisma/node_modules/@prisma/engines/query-engine-darwin)
Migration Engine     : migration-engine-cli e421996c87d5f3c8f7eeadd502d4ad402c89464d (at node_modules/prisma/node_modules/@prisma/engines/migration-engine-darwin)
Introspection Engine : introspection-core e421996c87d5f3c8f7eeadd502d4ad402c89464d (at node_modules/prisma/node_modules/@prisma/engines/introspection-engine-darwin)
Format Binary        : prisma-fmt e421996c87d5f3c8f7eeadd502d4ad402c89464d (at node_modules/prisma/node_modules/@prisma/engines/prisma-fmt-darwin)
Default Engines Hash : e421996c87d5f3c8f7eeadd502d4ad402c89464d
Studio               : 0.371.0
```

**Comment by P4sca1** [comment]
I upgraded to prisma `2.23.0` and now have a production instance running in debug mode with the latest version.
When the issue occurs again, I will let you know.

**Comment by pantharshit00** [comment]
Hey @P4sca1 

I will wait for your reply here :) 


Looks like something which won't be easy to reproduce but we will figure it out.

**Comment by P4sca1** [comment]
I did not yet see the issue again after upgrading. But this might also be related to the fact that we optimized our daily database backup job to no longer eat up all the memory and make the database server unresponsive for a few minutes in the night.

I believe what happened is that prisma timed out fetching connections because the datbase server was overwhelmed by the backup job. However for some reason after the database server got responsive again, prisma never recovered.

**Comment by pantharshit00** [comment]
I am going to close this. Without a reproduction this is unactionable.

**Comment by janpio** [comment]
Can you give us some additional information about that backup job @P4sca1? We might want to try to reproduce this as it is probably a quite common thing to happen - and Prisma should work well with that during that time.

**Comment by P4sca1** [comment]
There is a backup job running every night, which generates a database dump using `mysqldump` and then uploads it along with other files to a backup server. This uses `restic`, which is known to have a very high memory consumption during backups: https://github.com/restic/restic/issues/1988
This caused the server to have no free memory during the time of the backup. I can't say for sure, but my guess is that that was causing the Prisma connection timeouts and for some reason Prisma was not able to fetch new connections after the backup was done.

**Comment by janpio** [comment]
Ok, so filling up memory while running queries would be a viable reproduction attempt here.

**Comment by P4sca1** [comment]
The problem occured again and after further digging into metrics and logs, I am quite sure that this is not an issue with prisma, but with my web server running out of memory.

**Comment by janpio** [comment]
Thanks so much for the update @P4sca1. 

Prisma of course has a direct effect via our memory usage, so we will definitely investigate at some point in the future and find ways to manage the memory usage better. Right now it seems to be more of a uncommon problem, so we will probably prioritize other issues like features or bugs. But in the future we will be able to look at this issue again and take it into consideration (added some labels to make sure we actually find it again).

**Comment by P4sca1** [comment]
The reason the server ran out of memory was again restic. I never had critical problems with prismas memory usage.
I decided to upscale the machine resources to have more memory. This should hopefully get rid of the issue.
Thank you for your help in trying to find a solution! :)

**Comment by janpio** [comment]
Backups are hard... Hope you get that fixed some time somehow. Until then, have fun with Prisma.

**Comment by P4sca1** [comment]
Sadly, the issue started to arise again today at 08:23 EST. This time there was no backup running and the machine resources were not maxed out.

Web server:
![Bildschirmfoto 2021-06-11 um 09 44 21](https://user-images.githubusercontent.com/10992664/121650198-9bf0f900-ca99-11eb-8a0a-b2cfd2b93897.png)

Database server:
![Bildschirmfoto 2021-06-11 um 09 46 07](https://user-images.githubusercontent.com/10992664/121650474-e4101b80-ca99-11eb-90ba-8540bebaa33a.png)

I can login to the database and issue the `SHOW PROCESSLIST` command without a problem. It shows that no (0) connections are established from my prisma application.
Still the prisma application throws `Timed out fetching a connection from the pool`. 
![Bildschirmfoto 2021-06-11 um 09 50 18](https://user-images.githubusercontent.com/10992664/121650991-6f89ac80-ca9a-11eb-8084-35e6b4977f5e.png)
This is the amount of times prisma logged `Fetched a connection from the pool`:
![Bildschirmfoto 2021-06-11 um 09 51 05](https://user-images.githubusercontent.com/10992664/121651104-8c25e480-ca9a-11eb-980a-20ab480ff06f.png)

I still can't tell what's causing this and it seems to be really hard to debug. Would love to provide more info, but that's all I have for now.

I would expect Prisma to be able to fetch a connection when there are 0 connections currently open to the database. My guess is that there are some kind of broken connections in the Prisma pool, which prevent new connections from being made. But that's just a random thought without having detailed knowledge about the Prisma architecture.

**Comment by P4sca1** [comment]
As before, simply restarting the Prisma application resolved the problem.

**Comment by janpio** [comment]
Meh. What is your current `connect_timeout` and `pool_timeout` configuration when this happened? Do you by chance have some logs from the app using Prisma from when this happened?

**Comment by P4sca1** [comment]
> Meh. What is your current `connect_timeout` and `pool_timeout` configuration when this happened? Do you by chance have some logs from the app using Prisma from when this happened?

I have not changed both values, so `connect_timeout` should be 5 and `pool_timeout` should have a value of 10.

I was wrong when I said that the issue has nothing to do with the backup job. The graph I posted only showed the last hour, even though the last 12 hours of logs were processed.
Here is a correct graph counting the amount of "Timed out fetching..." logs from the last 12 hours.
<img width="1705" alt="Bildschirmfoto 2021-06-12 um 09 53 08" src="https://user-images.githubusercontent.com/10992664/121769458-ff475d80-cb63-11eb-87c6-113ba0fd9d3a.png">
The first spike is at 3am, then at 4am and 5 am (the hourly backup).

Even after load is back to normal at the database server, Prisma keeps timing out trying to fetch a connection.
A possible reproduction might be to put the database server under load (cpu and memory using some stress tester) until a timeout occurs in the Prisma app and then remove the load from the database server and see if Prisma is still timing out.

**Comment by TLadd** [comment]
I've been experiencing this as well. We get a temporary spike in requests to our app servers that increases memory and cpu near 100% for a short time. Prisma then seems to get stuck in a state of every query failing to fetch a connection and never recovers, even as cpu and memory usage drop. Restarting the servers returns things to normal. Wondering if https://github.com/prisma/prisma/issues/9584 is related since both seem to involve the client connections getting stuck in particular failure states.

**Comment by janpio** [comment]
That's a very good pointer @TLadd, and very possible. #9584 is on our list to reproduce very soon (And sounds much easier to reproduce than your other, spike related case), so hopefully we understand and fix that one soon - and might possibly also get rid of this one along the way.

**Comment by henrikuper** [comment]
Any update on this issue?

We experiencing the same. We have 2 prod servers that connect to 1 DB server. 
Randomly one of the serves get
```
Timed out fetching a new connection from the connection pool. More info: http://pris.ly/d/connection-pool (Current connection pool timeout: 10, connection limit: 5)
{"level":50,"time":1641985745165,"pid":143287,"hostname":"wf-prod-1","code":"P2024","clientVersion":"3.6.0","meta":{"connection_limit":5,"timeout":10}
```
After getting the error, every further request fails. And the prismaClient does not seem to recover at any time. 

The other server remains stable. 

A restart of the affected server solves the issue directly. But it can not be the solution to restart the server each time.

**Comment by janpio** [comment]
We actually have a reproduction now (essentially load spike or resource constraint with specific parameters) and are working on understanding the problem (non surprisingly something with the connection pool) and hopefully fixing it (no ETA - but if you are able to somehow reproduce this, we are looking for testers of our fixes of course).

Some related issues:
https://github.com/prisma/prisma/issues/11868
https://github.com/prisma/prisma-private/issues/160 (private with reproduction)

**Comment by hankewyczz** [comment]
We've been experiencing a similar issue (`v3.9.2`); after a day or two, all connections are rejected with the `Timed out fetching new connection` issue. The pool never recovers; a restart is necessary. Meanwhile, the actual underlying PSQL connections are all unused and available for queries.

Our issue may more closely match #11868, since there is no noticeable resource spike leading to the connection issue. We do execute a lot of queries in parallel, which seems to match the described behavior in that ticket. We'll try serializing the queries and seeing if that changes anything.

**Comment by janpio** [comment]
As mentioned above we are working on fixing this.

In the meantime, here is a possible temporary workaround you can try: 
Prisma offers a Client method to "disconnect" the current Engine from the Client: `prisma.$disconnect()`. By calling that method, we had good results in having Prisma Client internally "restart" the connection pool which allows you to avoid the restart of your application or server. 

We realize that this is not a long term solution for the problem of course, but hope it can work as a temporary workaround to unblock you. We'll update this issue when we are closer to a real solution.

**Comment by sreuter** [comment]
@garrensmith Not sure if https://github.com/prisma/prisma/pull/12066 fixes this as well?

**Comment by garrensmith** [comment]
This is fixed. The full information can be found here https://github.com/importcjj/mobc/issues/63
This will be out in the next release. You can test it now with `prisma@dev`

**Comment by w8ze-devel** [comment]
Hello,

We are experiencing the same problem as described in this issue.

- A default connection pool size based on the number of number of CPUs (default value is 21 in my case)
- No spike in CPU or memory when the problem occurs
- A restart of prisma fixed the issue

The application is used in production and there are not too much workloads but we can't figure out :

- why this issue occured 
- how to detect it with a supervision tool or something like this
- how to fixed it

We see that there's something involving mobc but don't know how to implement it.

The application is a backend API developed with NestJS.

Could someone give us a hint please ?

Thanks

**Comment by janpio** [comment]
Please open a new issue @w8ze-devel and provide all the information the bug issue template asks for. We are happy to jump into that and try to figure out - but we usually can not do that in closes issues where x other people are pinged on each response, I hope you understand. Thanks.

**Comment by 1mt4y** [comment]
I encountered this error while using a VPN (specifically ProtonVPN). It seems that either Vercel is blocking the VPN's IP, or there's an issue when using a VPN locally.

The problem was resolved for me by simply disconnecting from the VPN.

**Comment by AmrViernes** [comment]
Still have the same error we need to figure a workaround

**Comment by bpobocha** [comment]
same here, @w8ze-devel did you find any solution?

**Comment by ohthatsjhall** [comment]
I'm experiencing the same. Hoping to see a solution soon.

**Comment by janpio** [comment]
Everyone posting here: Please open a new issue or discussion and provide all the information the template asks for. We will not be able to help otherwise, as this is not a single bug or problem, but dependent on your individual situation.
